### ISSUE TITLE
Async log4j2 log events are not counted accurately

### ISSUE THREAD

**Issue by bol-com-pschmitz** [micrometer-metrics/micrometer#2176]
Micrometer provides the `Log4j2Metrics` meter binder to add metrics for Log4j2 log events. Unfortunately the implementation is flawed. It is implemented as a filter, but filters can be invoked multiple times and must therefore not have any side effects.

In practice the filter _is_ invoked multiple times by Log4j2 in many circumstances (e.g. in the presence of multiple and/or async loggers). The result is that the counter is increased multiple times for each single log event and the counter values become too high.

**Comment by bol-com-pschmitz** [comment]
Note that the issue is not that the filter is added multiple times. The same filter is invoked multiple times. You can for example see this is in `AsyncLoggingConfig.log(LogEvent, LoggerConfigPredicate)`. This first calls `super.log(...)`, which invokes the filter, and then it calls `logToAsyncDelegate(...)`, which again invokes the filter.

**Comment by schlangz** [comment]
So this only applies to async loggers?

**Comment by shakuzen** [comment]
@bol-com-pschmitz thank you for the report. Just so we're explicit about the scenario and that a fix properly addresses it, would you be able to write a test demonstrating it like other tests in our [Log4j2MetricsTest](https://github.com/micrometer-metrics/micrometer/blob/0c9aeef87bcb4f1fd1e89f6de7b63e8eba5af938/micrometer-core/src/test/java/io/micrometer/core/instrument/binder/logging/Log4j2MetricsTest.java)?

**Comment by bol-com-pschmitz** [comment]
> So this only applies to async loggers?

I'm not sure, possibly. That's how it occurred in my case. It doesn't matter though; `Log4j2Metrics` should not be implemented as a filter. Filters must always be able to be invoked multiple times and must never have side effects (such as incrementing metrics).

**Comment by bol-com-pschmitz** [comment]
> @bol-com-pschmitz would you be able to write a test demonstrating it like other tests in our Log4j2MetricsTest?

I'll take a look.

**Comment by dennysfredericci** [comment]
Hi all, hello from Belgium! :)

I just create a test for this issue and indeed the filter have been called multiple times for same event.

**Comment by bol-com-pschmitz** [comment]
Thanks @dennysfredericci ! I was about to start doing that.

**Comment by dennysfredericci** [comment]
There is a method isEndOfBatch on Log4jEventLog which can be used to see if this is the "last" call.

`
    /**
     * Returns {@code true} if this event is the last one in a batch, {@code false} otherwise. Used by asynchronous
     * Loggers and Appenders to signal to buffered downstream components when to flush to disk, as a more efficient
     * alternative to the {@code immediateFlush=true} configuration.
     *
     * @return whether this event is the last one in a batch.
     */
    // see also LOG4J2-164
    boolean isEndOfBatch();
`

I changed the implementation to test this idea, but it breaks all other tests.

As an filter can be called multiple times the only solution for this issue is a custom appender instead of a filter.

**Comment by dennysfredericci** [comment]
Hi Guys,

When we have an async logger configured a call to log.info or any other level triggers 2 times the filter, i found a way to use the isEndOfBatch attribute from LogEvent to check if this is the last call and then metrics counter will be incremented only once instead of multiple times.

I also create a simple spring-boot project to test it locally. You can see it here: https://github.com/dennysfredericci/springboot-micrometer-issue-2183

**Comment by dennysfredericci** [comment]
Hi @shakuzen 

Just noticed now, I did the PR for master branch should I change it to 1.1.X branch?

**Comment by shakuzen** [comment]
Resolved by #2183

**Comment by shakuzen** [comment]
@bol-com-pschmitz would you confirm if the issue is fixed for you? The fix is available in versions 1.1.16, 1.3.11, and it will be in the upcoming 1.5.3 release (you can try the latest snapshot version 1.5.3-SNAPSHOT for now).

**Comment by bol-com-pschmitz** [comment]
> @bol-com-pschmitz would you confirm if the issue is fixed for you? The fix is available in versions 1.1.16, 1.3.11, and it will be in the upcoming 1.5.3 release (you can try the latest snapshot version 1.5.3-SNAPSHOT for now).

I'm testing it. At first glance it looks like now it might be undercounting. I'm not sure what constitutes a "batch", but if it can result in more than one log entry than this is not going to work.

**Comment by shakuzen** [comment]
@bol-com-pschmitz Thanks for testing. Let us know if there's another issue. If we can get a unit test to reproduce, that'll be best so we can make sure it gets fixed properly and stays fixed.

**Comment by bol-com-pschmitz** [comment]
Unfortunately this approach does not work. It counts batches of log events, not log events. This small change to the unit test added by @dennysfredericci demonstrates it:

```java
        assertThat(registry.get("log4j2.events").tags("level", "info").counter().count()).isEqualTo(0);
        logger.info("Hello, world!");
        logger.info("Hello, world!");
        logger.info("Hello, world!");
        assertThat(registry.get("log4j2.events").tags("level", "info").counter().count()).isEqualTo(3);
```

**Comment by bol-com-pschmitz** [comment]
This means the issue has not been fixed and should be reopened.

**Comment by dennysfredericci** [comment]
Indeed I can see it as well :disappointed:

**Comment by bol-com-pschmitz** [comment]
Sorry @dennysfredericci . I appreciate your effort! But I still think that this should not be implemented as a filter at all. Filters can be invoked any number of times and should not have side effects such as incrementing counters.

**Comment by shakuzen** [comment]
I'm no log4j2 expert, so I'm hoping we can get some help on a solution for this from someone more knowledgeable than me on this area. I've marked the issue as `help wanted`.

**Comment by dennysfredericci** [comment]
So, let's became a log4j2 expert :smile:

I just started a thread on[ log4j2 discussion list](https://lists.apache.org/thread.html/r404429bc0d2a64177a93da0664e5738db321178db761d10e89f934a3%40%3Cdev.logging.apache.org%3E) to see if is possible to use a filter for that use case.

**Comment by dennysfredericci** [comment]
> Sorry @dennysfredericci . I appreciate your effort! But I still think that this should not be implemented as a filter at all. Filters can be invoked any number of times and should not have side effects such as incrementing counters.

Hi @bol-com-pschmitz  this seems an unexpected behavior, I got the message below on log4j mail list

> "... Filters that are attached to Loggers, Appender-Refs and Appenders should only be called once ..."

I just included a [reconfiguration](https://github.com/dennysfredericci/micrometer/blob/bugfix/2176-log4j2-async-counter/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/logging/Log4j2Metrics.java#L101) call and now the filter is called just once, but I am getting a new exception.

```
2020-08-11 22:24:28,784 Test worker ERROR Could not register mbeans java.lang.NullPointerException
	at org.apache.logging.log4j.core.async.AsyncLoggerConfigDisruptor.createRingBufferAdmin(AsyncLoggerConfigDisruptor.java:417)
	at org.apache.logging.log4j.core.async.AsyncLoggerConfig.createRingBufferAdmin(AsyncLoggerConfig.java:199)
	at org.apache.logging.log4j.core.jmx.Server.registerLoggerConfigs(Server.java:366)
	at org.apache.logging.log4j.core.jmx.Server.reregisterMBeansAfterReconfigure(Server.java:186)
	at org.apache.logging.log4j.core.jmx.Server.reregisterMBeansAfterReconfigure(Server.java:141)
	at org.apache.logging.log4j.core.LoggerContext.setConfiguration(LoggerContext.java:629)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:712)
	at org.apache.logging.log4j.core.config.Configurator.reconfigure(Configurator.java:239)
	at io.micrometer.core.instrument.binder.logging.Log4j2Metrics.bindTo(Log4j2Metrics.java:101)
```

In summary, this can be an issue in how we add the filters to log4j configuration.

**Comment by dennysfredericci** [comment]
As is not possible to use log4j filters to count log events properly I [requested a statistics feature](https://issues.apache.org/jira/browse/LOG4J2-2937) for the log4j team.

With some luck, we can have these counters direct from log4j API.

**Comment by shakuzen** [comment]
https://github.com/apache/logging-log4j2/issues/1550 has been resolved and is marked for inclusion in log4j2 2.20.1. Once that is released and we've upgraded, we'll add a test along the lines of https://github.com/micrometer-metrics/micrometer/issues/2176#issuecomment-669265851 to ensure the issue no longer affects us. Hopefully then we can close this issue.

**Comment by jonatan-ivanov** [comment]
It seems log4j2 `2.20.1` was not released, the latest patch release on the `2.20.x` line is  [`2.20.0`](https://logging.apache.org/log4j/2.x/release-notes.html#release-notes-2-20-0) (release date: 2023-02-17) and the fix (https://github.com/apache/logging-log4j2/issues/1550) was released in [`2.21.0`](https://logging.apache.org/log4j/2.x/release-notes.html#release-notes-2-21-0).

Our `main` branch was upgraded to this version and I eliminated the previous `isEndOfBatch` check in https://github.com/micrometer-metrics/micrometer/pull/4253 but I have a few questions:
1. Do we need to handle any way if an earlier log4j2 version is on the classpath? I think we don't need to do anything since the behavior was broken one way or the other.
2. Do we want to back-port this to earlier versions too, like 1.9.x? I think so since log4j2 is an optional dependency for us.
3. Could you please check if my changes and what I wrote down in the commit message seems ok? (see https://github.com/micrometer-metrics/micrometer/pull/4253/commits/61b3074459b825111e9b3064e03594c7f8f80c7a and https://github.com/micrometer-metrics/micrometer/pull/4253)
